xs/src/libslic3r/Polyline.cpp

```diff
@@ -251,4 +252,105 @@ ThickPolyline::reverse()
     std::swap(this->endpoints.first, this->endpoints.second);
 }
 
+void
+concatThickPolylines(ThickPolylines& pp) {
+    bool changes = true;
+    while (changes){
+        changes = false;
+        //concat polyline if only 2 polyline at a point
+        for (size_t i = 0; i < pp.size(); ++i) {
+            ThickPolyline *polyline = &pp[i];
+            if (polyline->first_point().coincides_with(polyline->last_point())) {
+                polyline->endpoints.first = false;
+                polyline->endpoints.second = false;
+                continue;
+            }
+
+            size_t id_candidate_first_point = -1;
+            size_t id_candidate_last_point = -1;
+            size_t nbCandidate_first_point = 0;
+            size_t nbCandidate_last_point = 0;
+            // find another polyline starting here
+            for (size_t j = 0; j < pp.size(); ++j) {
```

Comment:
Starting from, `i` may be enough ?

Reply:
No

Reply:
I'm overlooking something .. 
IMO all tested `i` polylines should be extended as far as possible (on points that join exactly 2 polylines). So there shall be no valid candidate with `j<i`.
For same reason, `changes` should affect only current `i` polyline, all polylines before `i` are already processed.

Reply:
I had to revert from your proposal to this version because it doesn't work.
It's because it can only merge 2 lines at each iteration of i. After that, we may be able to merge with other lines, so "other" must have the possibility to find us, even if our idx is lower.

Reply:
(I may be wrong..)
Yes, but all 'others' are `j > i` - merging is commutative, the result (test) is same if merging i+j or j+i.
All that is necessary is to rescan `i` again if it was merged (`i--; continue;` or better increase `i` outside of `for`, only when no candidate was found)

Reply:
(sorry, It is likely that I overlooked something. Happens all the the time. It's just hard let it go until I find the problem)

Reply:
I can't find the code i made to test it with j=i+1. If you can make it works, i will happily merge with your modification.